cli/lib/nf3.mjs

```diff
@@ -204,47 +206,72 @@ class Nf3 {
     return axios.get(`${this.optimistBaseUrl}/block/make-now`);
   }
 
+  async estimateGas(contractAddress, unsignedTransaction) {
+    let gasLimit;
+    try {
+      // eslint-disable-next-line no-await-in-loop
+      gasLimit = await this.web3.eth.estimateGas({
+        from: this.ethereumAddress,
+        to: contractAddress,
+        data: unsignedTransaction,
+      });
+    } catch (error) {
+      logger.warn(`estimateGas failed. Falling back to constant value`);
+      gasLimit = GAS; // backup if estimateGas failed
+    }
+    return Math.ceil(Number(gasLimit) * GAS_MULTIPLIER); // 50% seems a more than reasonable buffer.
+  }
+
+  async estimateGasPrice() {
+    let proposedGasPrice;
+    try {
+      // Call the endpoint to estimate the gas fee.
+      const res = (await axios.get(GAS_ESTIMATE_ENDPOINT)).data.result;
```

Comment:
Why are we using a lamba function? Can’t the call to Etherscan be embedded directly?

Reply:
Mostly for ease-of-use. The etherscan call requires an API_KEY that needs to be kept private. For the `sdk` we could extract this secret information from an environment variable; however, the wallet also needs this and there is no way to keep this information secret in the browser.

So we put it behind a gateway and lambda (where the API_KEY is stored in the secrets manager).